src/iop/channelmixerrgb.c

```diff
@@ -0,0 +1,2468 @@
+    }
+  }
+  else
+  {
+    // error is too big to use a CCT-based model, we fall back to a custom/freestyle chroma selection for the illuminant
+    if(illuminant) *illuminant = DT_ILLUMINANT_CUSTOM;
+
+    // CAT16 is less accurate but more robust for non-daylight (produces fewer out-of-gamut colors)
+    if(adaptation) *adaptation = DT_ADAPTATION_CAT16;
+  }
+}
+
+
+void process(struct dt_iop_module_t *self, dt_dev_pixelpipe_iop_t *piece, const void *const restrict ivoid,
+             void *const restrict ovoid, const dt_iop_roi_t *const roi_in, const dt_iop_roi_t *const roi_out)
+{
+  const dt_iop_channelmixer_rbg_data_t *data = (dt_iop_channelmixer_rbg_data_t *)piece->data;
+  const dt_iop_order_iccprofile_info_t *const work_profile = dt_ioppr_get_pipe_work_profile_info(piece->pipe);
+  dt_iop_channelmixer_rgb_gui_data_t *g = (dt_iop_channelmixer_rgb_gui_data_t *)self->gui_data;
+
+  float DT_ALIGNED_ARRAY RGB_to_XYZ[3][4];
+  float DT_ALIGNED_ARRAY XYZ_to_RGB[3][4];
+
+  // repack the matrices as flat AVX2-compliant matrice
+  if(work_profile)
+  {
+    // work profile can't be fetched in commit_params since it is not yet initialised
+    repack_3x3_to_3xSSE(work_profile->matrix_in, RGB_to_XYZ);
+    repack_3x3_to_3xSSE(work_profile->matrix_out, XYZ_to_RGB);
+  }
+
+  assert(piece->colors == 4);
+  const size_t ch = 4;
+
+  const float *const restrict in = (const float *const restrict)ivoid;
+  float *const restrict out = (float *const restrict)ovoid;
+
+  // auto-detect WB upon request
+  if(self->dev->gui_attached && g)
+  {
+    if(g->auto_detect_illuminant && !darktable.gui->reset)
+    {
+      if(piece->pipe->type != DT_DEV_PIXELPIPE_FULL)
+      {
+        // detection on full image only
+        dt_simd_memcpy(in, out, roi_in->width * roi_in->height * ch);
+        return;
+      }
+
+      float XYZ[4] = { 0.f };
+      auto_detect_WB(in, g->wb_model, roi_in->width, roi_in->height, ch, RGB_to_XYZ, XYZ);
+
+      ++darktable.gui->reset;
+      dt_iop_channelmixer_rgb_params_t *p = (dt_iop_channelmixer_rgb_params_t *)self->params;
+      p->x = XYZ[0];
+      p->y = XYZ[1];
+
+      check_if_close_to_daylight(p->x, p->y, &p->temperature, &p->illuminant, &p->adaptation);
+
+      dt_bauhaus_slider_set(g->temperature, p->temperature);
+      dt_bauhaus_combobox_set(g->illuminant, p->illuminant);
+      dt_bauhaus_combobox_set(g->adaptation, p->adaptation);
+
+      float xyY[3] = { p->x, p->y, 1.f };
+      float Lch[3];
+      dt_xyY_to_Lch(xyY, Lch);
+      dt_bauhaus_slider_set(g->illum_x, Lch[2] / M_PI * 180.f);
+      dt_bauhaus_slider_set(g->illum_y, Lch[1]);
+
+      update_illuminants(self);
+      update_approx_cct(self);
+      update_illuminant_color(self);
+      g->auto_detect_illuminant = FALSE;
+
+      --darktable.gui->reset;
+
+      dt_control_log(_("auto-detection of white balance completed"));
+
+      dt_dev_add_history_item(darktable.develop, self, TRUE);
+      return;
+    }
+  }
```

Comment:
All the GUI stuff in this section should be done in gui_update().

Reply:
Are you sure `gui_update()` is triggered after `process()` exits ? `auto_detect_WB(in, g->wb_model, roi_in->width, roi_in->height, ch, RGB_to_XYZ, XYZ);` writes the illuminant found in `XYZ`, but then we need to commit that to the parameters and update temperature and so on, then finally the GUI.

Reply:
I'm not sure about update indeed, but I'm sure no Gtk code should be used in process() which is run in a different thread.

Reply:
The best solution I suppose is to remove those two entries from the combo and add buttons that act like color-picker to do the detection.

Reply:
But that won't change a thing if we put the detection in buttons, we will still need to run an algo on the pixels, wait for the processing to finish, get the result then output the value to GUI and update the widgets accordingly. Such things are done already in ashift, exposure and tone equalizer, although they use private `dt_pthread_mutex_unlock(&g->lock);`.

Reply:
Fixed